File: apps/web/src/lib/action-router/logger.ts
/**
 * Action Router - Logger
 *
 * Log all action executions to the database for audit trail.
 */

import { supabase } from '@/lib/supabaseClient';
import type { ActionLogEntry, ActionStatus } from './types';

// ============================================================================
// SENSITIVE FIELD PATTERNS
// ============================================================================

const SENSITIVE_PATTERNS = [
  /password/i,
  /secret/i,
  /token/i,
  /api_key/i,
  /apikey/i,
  /credential/i,
  /private_key/i,
  /auth/i,
];

// ============================================================================
// SANITIZATION
// ============================================================================

/**
 * Sanitize payload by redacting sensitive fields
 *
 * @param payload - Payload to sanitize
 * @returns Sanitized payload
 */
export function sanitizePayload(
  payload: Record<string, unknown>
): Record<string, unknown> {
  const sanitized: Record<string, unknown> = {};

  for (const [key, value] of Object.entries(payload)) {
    // Check if key matches sensitive pattern
    const isSensitive = SENSITIVE_PATTERNS.some((pattern) => pattern.test(key));

    if (isSensitive) {
      sanitized[key] = '[REDACTED]';
    } else if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
      // Recursively sanitize nested objects
      sanitized[key] = sanitizePayload(value as Record<string, unknown>);
    } else {
      sanitized[key] = value;
    }
  }

  return sanitized;
}

// ============================================================================
// ACTION LOGGING
// ============================================================================

/**
 * Log an action execution to the database
 *
 * @param entry - Log entry data
 * @returns Log entry ID or null on error
 */
export async function logAction(entry: {
  actionId: string;
  actionLabel: string;
  yachtId: string;
  userId: string;
  payload: Record<string, unknown>;
  status: ActionStatus;
  result?: Record<string, unknown>;
  errorMessage?: string;
  durationMs?: number;
}): Promise<string | null> {
  try {
    // Sanitize payload and result
    const sanitizedPayload = sanitizePayload(entry.payload);
    const sanitizedResult = entry.result
      ? sanitizePayload(entry.result)
      : undefined;

    const logEntry: ActionLogEntry = {
      yacht_id: entry.yachtId,
      user_id: entry.userId,
      action_id: entry.actionId,
      action_label: entry.actionLabel,
      payload: sanitizedPayload,
      status: entry.status,
      result: sanitizedResult,
      error_message: entry.errorMessage,
      duration_ms: entry.durationMs,
    };

    const { data, error } = await supabase
      .from('action_executions')
      .insert(logEntry)
      .select('id')
      .single();

    if (error) {
      console.error('Failed to log action:', error);
      return null;
    }

    return data.id;
  } catch (error) {
    console.error('Failed to log action:', error);
    return null;
  }
}

// ============================================================================
// ACTION STATISTICS
// ============================================================================

/**
 * Get action execution statistics for a yacht
 *
 * @param yachtId - Yacht ID
 * @param hours - Time period in hours (default 24)
 * @returns Action statistics
 */
export async function getActionStats(
  yachtId: string,
  hours: number = 24
): Promise<{
  total: number;
  success: number;
  error: number;
  byAction: Record<string, { total: number; success: number }>;
  avgDurationMs: number;
}> {
  try {
    const since = new Date(Date.now() - hours * 60 * 60 * 1000).toISOString();

    const { data, error } = await supabase
      .from('action_executions')
      .select('action_id, status, duration_ms')
      .eq('yacht_id', yachtId)
      .gte('created_at', since);

    if (error) {
      throw error;
    }

    const stats = {
      total: 0,
      success: 0,
      error: 0,
      byAction: {} as Record<string, { total: number; success: number }>,
      avgDurationMs: 0,
    };

    let totalDuration = 0;
    let durationCount = 0;

    for (const row of data || []) {
      stats.total++;

      if (row.status === 'success') {
        stats.success++;
      } else {
        stats.error++;
      }

      // Track by action
      if (!stats.byAction[row.action_id]) {
        stats.byAction[row.action_id] = { total: 0, success: 0 };
      }
      stats.byAction[row.action_id].total++;
      if (row.status === 'success') {
        stats.byAction[row.action_id].success++;
      }

      // Track duration
      if (row.duration_ms) {
        totalDuration += row.duration_ms;
        durationCount++;
      }
    }

    stats.avgDurationMs =
      durationCount > 0 ? Math.round(totalDuration / durationCount) : 0;

    return stats;
  } catch (error) {
    console.error('Failed to get action stats:', error);
    return {
      total: 0,
      success: 0,
      error: 0,
      byAction: {},
      avgDurationMs: 0,
    };
  }
}

// ============================================================================
// RECENT ACTIONS
// ============================================================================

/**
 * Get recent action executions for a yacht
 *
 * @param yachtId - Yacht ID
 * @param limit - Maximum number of results (default 50)
 * @returns Recent action log entries
 */
export async function getRecentActions(
  yachtId: string,
  limit: number = 50
): Promise<ActionLogEntry[]> {
  try {
    const { data, error } = await supabase
      .from('action_executions')
      .select('*')
      .eq('yacht_id', yachtId)
      .order('created_at', { ascending: false })
      .limit(limit);

    if (error) {
      throw error;
    }

    return (data || []) as ActionLogEntry[];
  } catch (error) {
    console.error('Failed to get recent actions:', error);
    return [];
  }
}

// ============================================================================
// ACTION HISTORY FOR ENTITY
// ============================================================================

/**
 * Get action history for a specific entity
 *
 * @param yachtId - Yacht ID
 * @param entityType - Type of entity (equipment_id, work_order_id, etc.)
 * @param entityId - Entity ID
 * @param limit - Maximum number of results (default 20)
 * @returns Action log entries for entity
 */
export async function getEntityActionHistory(
  yachtId: string,
  entityType: string,
  entityId: string,
  limit: number = 20
): Promise<ActionLogEntry[]> {
  try {
    // Query for actions where payload contains the entity
    const { data, error } = await supabase
      .from('action_executions')
      .select('*')
      .eq('yacht_id', yachtId)
      .contains('payload', { [entityType]: entityId })
      .order('created_at', { ascending: false })
      .limit(limit);

    if (error) {
      throw error;
    }

    return (data || []) as ActionLogEntry[];
  } catch (error) {
    console.error('Failed to get entity action history:', error);
    return [];
  }
}

File: apps/api/action_router/logger.py
"""
Action Logger

Logs all action executions to Supabase for audit trail and analytics.
"""

from typing import Dict, Any, Optional
import os
from datetime import datetime
from supabase import Client
import json

# Import centralized Supabase client factory
from integrations.supabase import get_supabase_client


def sanitize_payload(payload: Dict[str, Any]) -> Dict[str, Any]:
    """
    Sanitize sensitive data from payload before logging.

    Removes or masks sensitive fields like passwords, tokens, etc.
    """
    sensitive_fields = [
        "password",
        "token",
        "secret",
        "api_key",
        "access_token",
        "refresh_token",
    ]

    sanitized = payload.copy()

    for key in list(sanitized.keys()):
        # Check if key contains sensitive field name
        if any(sensitive in key.lower() for sensitive in sensitive_fields):
            sanitized[key] = "***REDACTED***"

    return sanitized


async def log_action(
    action_id: str,
    action_label: str,
    yacht_id: str,
    user_id: str,
    payload: Dict[str, Any],
    status: str,
    result: Optional[Dict[str, Any]] = None,
    error_message: Optional[str] = None,
    duration_ms: Optional[int] = None,
    execution_id: Optional[str] = None,
) -> None:
    """
    Log action execution to Supabase.

    Args:
        action_id: ID of the action
        action_label: Human-readable label
        yacht_id: UUID of the yacht
        user_id: UUID of the user
        payload: Action payload (will be sanitized)
        status: 'success' or 'error'
        result: Result data (optional)
        error_message: Error message if status='error'
        duration_ms: Execution duration in milliseconds
        execution_id: Unique ID for this execution
    """
    try:
        supabase = get_supabase_client()

        # Sanitize payload
        sanitized_payload = sanitize_payload(payload)

        # Prepare log entry
        log_entry = {
            "action_id": action_id,
            "action_label": action_label,
            "yacht_id": yacht_id,
            "user_id": user_id,
            "payload": json.dumps(sanitized_payload),
            "status": status,
            "timestamp": datetime.utcnow().isoformat(),
        }

        if execution_id:
            log_entry["execution_id"] = execution_id

        # Add optional fields
        if result:
            # Don't log full result if it's large, just summary
            if isinstance(result, dict):
                result_summary = {k: v for k, v in result.items() if k in ["id", "note_id", "work_order_id", "handover_id", "signed_url"]}
                log_entry["result"] = json.dumps(result_summary)
            else:
                log_entry["result"] = json.dumps(result)

        if error_message:
            log_entry["error_message"] = error_message[:500]  # Truncate long errors

        if duration_ms is not None:
            log_entry["duration_ms"] = duration_ms

        # Insert log
        supabase.table("action_logs").insert(log_entry).execute()

    except Exception as e:
        # Log errors should not break the action execution
        # Just print to stderr for debugging
        import sys
        print(f"WARNING: Failed to log action: {str(e)}", file=sys.stderr)


async def get_action_stats(yacht_id: str, hours: int = 24) -> Dict[str, Any]:
    """
    Get action execution statistics for a yacht.

    Args:
        yacht_id: UUID of the yacht
        hours: Number of hours to look back

    Returns:
        Dictionary with stats:
        - total_actions: Total number of actions
        - success_rate: Percentage of successful actions
        - actions_by_type: Count by action_id
        - recent_errors: List of recent errors
    """
    try:
        supabase = get_supabase_client()

        # Calculate timestamp for N hours ago
        from datetime import timedelta
        cutoff = (datetime.utcnow() - timedelta(hours=hours)).isoformat()

        # Get all logs for yacht in time range
        result = supabase.table("action_logs").select("*").eq(
            "yacht_id", yacht_id
        ).gte("timestamp", cutoff).execute()

        logs = result.data

        # Calculate stats
        total_actions = len(logs)
        successful_actions = len([log for log in logs if log["status"] == "success"])
        success_rate = (successful_actions / total_actions * 100) if total_actions > 0 else 0

        # Count by action type
        actions_by_type = {}
        for log in logs:
            action_id = log["action_id"]
            actions_by_type[action_id] = actions_by_type.get(action_id, 0) + 1

        # Get recent errors
        recent_errors = [
            {
                "action_id": log["action_id"],
                "error_message": log["error_message"],
                "timestamp": log["timestamp"],
            }
            for log in logs
            if log["status"] == "error"
        ][:10]  # Last 10 errors

        return {
            "total_actions": total_actions,
            "success_rate": round(success_rate, 2),
            "actions_by_type": actions_by_type,
            "recent_errors": recent_errors,
            "time_range_hours": hours,
        }

    except Exception as e:
        raise Exception(f"Failed to get action stats: {str(e)}")


__all__ = ["log_action", "get_action_stats", "sanitize_payload"]
